Microsoft.Toolkit.Uwp.UI.Controls.Primitives/ConstrainedPresenter/ConstrainedPresenter.cs

```diff
@@ -0,0 +1,119 @@
+                panel.InvalidateMeasure();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the scale for the width of the panel. Should be a value between 0-1.0. Default is 1.0.
+        /// </summary>
+        public double ScaleX
+        {
+            get { return (double)GetValue(ScaleXProperty); }
+            set { SetValue(ScaleXProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="ScaleX"/> property.
+        /// </summary>
+        public static readonly DependencyProperty ScaleXProperty =
+            DependencyProperty.Register(nameof(ScaleX), typeof(double), typeof(ConstrainedPresenter), new PropertyMetadata(1.0, ScalePropertyChanged));
+
+        /// <summary>
+        /// Gets or sets the scale for the height of the panel. Should be a value between 0-1.0. Default is 1.0.
+        /// </summary>
+        public double ScaleY
+        {
+            get { return (double)GetValue(ScaleYProperty); }
+            set { SetValue(ScaleYProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="ScaleY"/> property.
+        /// </summary>
+        public static readonly DependencyProperty ScaleYProperty =
+            DependencyProperty.Register(nameof(ScaleY), typeof(double), typeof(ConstrainedPresenter), new PropertyMetadata(1.0, ScalePropertyChanged));
+
+        private static void ScalePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if (d is ConstrainedPresenter panel)
+            {
+                panel.InvalidateMeasure();
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            return base.MeasureOverride(CalculateConstrainedSize(availableSize));
+        }
+
+        /// <inheritdoc/>
+        protected override Size ArrangeOverride(Size finalSize)
+        {
+            return base.ArrangeOverride(CalculateConstrainedSize(finalSize));
+        }
+
+        private Size CalculateConstrainedSize(Size initialSize)
+        {
+            var availableSize = new Size(initialSize.Width * ScaleX, initialSize.Height * ScaleY);
+
+            // If we don't have an Aspect Ratio, just return the scaled value.
+            if (ReadLocalValue(AspectRatioProperty) == DependencyProperty.UnsetValue)
+            {
+                return availableSize;
+            }
+
+            // Calculate the Aspect Ratio constraint based on the newly scaled size.
+            var currentAspect = availableSize.Width / availableSize.Height;
+            var desiredAspect = AspectRatio.Value;
+
+            if (currentAspect >= desiredAspect)
+            {
+                return new Size(availableSize.Height * desiredAspect, availableSize.Height);
+            }
+            else
+            {
+                return new Size(availableSize.Width, availableSize.Width / desiredAspect);
+            }
```

Comment:
Can we add tests for this? This feels like something that could easily break if not paid enough attention too.

Reply:
Yeah, I plan to add tests, it's in the TODO list. 🙂